File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const map = new Map()
    
    map.set(')', '(')
    map.set(']', '[')
    map.set('}', '{')
    
    const stack = []
    
    for (const c of s) {
        if (!map.has(c)) {
            stack.push(c)
        } else if (stack.length === 0 || stack.pop() !== map.get(c)) {
            return false
        }
    }
    
    return stack.length === 0
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> brackets = new HashMap<>();
        brackets.put(')', '(');
        brackets.put('}', '{');
        brackets.put(']', '[');
        
        Stack<Character> stack = new Stack<>();
        for (char c: s.toCharArray()) {
            if (!brackets.containsKey(c)) {
                stack.push(c);
            } else if (stack.isEmpty() || brackets.get(c) != stack.pop()) {
                return false;
            }
        }
        
        return stack.isEmpty();
    }
}